compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -985,7 +985,180 @@ OMR::Power::TreeEvaluator::vstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
+   TR::Node *firstChild = node->getFirstChild();
+   TR::DataType type = firstChild->getDataType().getVectorElementType();
+
+   TR_ASSERT_FATAL_WITH_NODE(node, firstChild->getDataType().getVectorLength() == TR::VectorLength128,
+                   "Only 128-bit vectors are supported %s", firstChild->getDataType().toString());
+
+   switch(type)
+     {
+     case TR::Int8:
+       return TR::TreeEvaluator::vreductionAddSubWordHelper(node,cg, TR::InstOpCode::vsum4sbs); //since the results will never overflow, it is safe to use a saturate sum instruction here
+     case TR::Int16:
+       return TR::TreeEvaluator::vreductionAddSubWordHelper(node,cg, TR::InstOpCode::vsum4shs); //since the results will never overflow, it is safe to use a saturate sum instruction here
+     case TR::Int32:
+       return TR::TreeEvaluator::vreductionAddWordHelper(node,cg, type);
+     case TR::Int64:
+       return TR::TreeEvaluator::vreductionAddDoubleWordHelper(node,cg, type);
+     case TR::Float:
+       return TR::TreeEvaluator::vreductionAddWordHelper(node,cg, type);
+     case TR::Double:
+       return TR::TreeEvaluator::vreductionAddDoubleWordHelper(node,cg, type);
+     default:
+       TR_ASSERT_FATAL(false, "unrecognized vector type %s\n", firstChild->getDataType().toString()); return NULL;
+     }
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddSubWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::InstOpCode::Mnemonic sumOp)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg = cg->allocateRegister();
+
+   TR::Register *tempRes = cg->allocateRegister(TR_VRF);
+   TR::Register *zeroReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(resReg);
+
+   //Evaluate sum across vector operand. Since the operands will be either 8 or 16 bytes long, and thus the results cannot overflow, it is safe to use saturate sum instructions here
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, zeroReg, 0);
+   generateTrg1Src2Instruction(cg, sumOp, node, tempRes, srcReg, zeroReg); //Sum across each word element
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vsumsws, node, tempRes, tempRes, zeroReg); //Sum word elements across entire vector operand
+
+   //Copy result into GPR
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9))
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrld, node, resReg, tempRes);
+   else
+   {
+      generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, tempRes, tempRes, zeroReg, 3); //move sum to upper doubleword element of tempRes
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrd, node, resReg, tempRes);
+   }
+
+   cg->stopUsingRegister(tempRes);
+   cg->stopUsingRegister(zeroReg);
+   cg->decReferenceCount(firstChild);
+
+   return resReg;
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::DataType type)
```

Comment:
Is there any chance for other .cpp files call `vreductionAddWordHelper()`?
If no, how about making it a static function in OMRTreeEvaluator.cpp instead of a member function of `OMR::Power::TreeEvaluator`?